pkg/providers/aws/provider_postgressnapshot.go

```diff
@@ -0,0 +1,244 @@
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	session, err := p.getAwsSession(ctx, postgres)
+
+	if err != nil {
+		errMsg := "failed to create AWS session"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	rdsSvc := rds.New(session)
+
+	return p.createPostgresSnapshot(ctx, snapshot, postgres, rdsSvc)
+}
+
+func (p *PostgresSnapshotProvider) DeletePostgresSnapshot(ctx context.Context, snapshot *v1alpha1.PostgresSnapshot, postgres *v1alpha1.Postgres) (croType.StatusMessage, error) {
+
+	session, err := p.getAwsSession(ctx, postgres)
+
+	if err != nil {
+		errMsg := "failed to create AWS session"
+		return croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	rdsSvc := rds.New(session)
+
+	return p.deletePostgresSnapshot(ctx, snapshot, postgres, rdsSvc)
+}
+
+func (p *PostgresSnapshotProvider) createPostgresSnapshot(ctx context.Context, snapshot *v1alpha1.PostgresSnapshot, postgres *v1alpha1.Postgres, rdsSvc rdsiface.RDSAPI) (*providers.PostgresSnapshotInstance, croType.StatusMessage, error) {
+	// generate snapshot name
+	snapshotName, err := BuildTimestampedInfraNameFromObjectCreation(ctx, p.client, snapshot.ObjectMeta, DefaultAwsIdentifierLength)
+	if err != nil {
+		errMsg := "failed to generate snapshot name"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// update cr with snapshot name
+	snapshot.Status.SnapshotID = snapshotName
+
+	if err = p.client.Status().Update(ctx, snapshot); err != nil {
+		errMsg := fmt.Sprintf("failed to update instance %s in namespace %s", snapshot.Name, snapshot.Namespace)
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// get instance name
+	instanceName, err := BuildInfraNameFromObject(ctx, p.client, postgres.ObjectMeta, DefaultAwsIdentifierLength)
+	if err != nil {
+		errMsg := "failed to get cluster name"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	foundSnapshot, err := p.findSnapshotInstance(rdsSvc, snapshotName)
+
+	if err != nil {
+		errMsg := "failed to describe snaphots in AWS"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// create snapshot of the rds instance
+	if foundSnapshot == nil {
+		// postgres instance has either just been created
+		// or is already backing up
+		if postgres.Status.Phase == croType.PhaseInProgress {
+			return nil, croType.StatusMessage("waiting for postgres instance to be available"), nil
+		}
+		// postgres instance is being deleted
+		// impossible to create a snapshot
+		if postgres.Status.Phase == croType.PhaseDeleteInProgress {
+			errMsg := "cannot create snapshot when instance deletion is in progress"
+			return nil, croType.StatusMessage(errMsg), errorUtil.New(errMsg)
+		}
+		p.logger.Info("creating rds snapshot")
+		_, err = rdsSvc.CreateDBSnapshot(&rds.CreateDBSnapshotInput{
```

Comment:
Should we be tagging these SnapShots on creation?

Reply:
Good point. I don't think the old logic tagged the snapshots either. I could add that here as part of this PR but I wonder would it make more sense to create a separate ticket for it?

Reply:
tut tut tut who wrote that old logic :rofl: 

if we can get it in in this PR I think we should as it should be a small change, the `CreateDBSnapshotInput` has a [tags](https://docs.aws.amazon.com/sdk-for-go/api/service/rds/#CreateDBSnapshotInput) field

But of course elasticache does not have a [tags](https://docs.aws.amazon.com/sdk-for-go/api/service/elasticache/#CreateSnapshotInput) field. :thinking: 

Ill leave it up to you, I dont want to hold up this PR if it will take too much time, another follow up ticket might be worth it.

Reply:
Okay I'll definitely get it in there for Postgres and I'll try get it in for Redis 😄

Reply:
I started working on this but then I discovered that snapshots are actually tagged within the respective postgres and redis providers.

Postgres: https://github.com/integr8ly/cloud-resource-operator/blob/master/pkg/providers/aws/provider_postgres.go#L367-L391
Redis https://github.com/integr8ly/cloud-resource-operator/blob/master/pkg/providers/aws/provider_redis.go#L377-L397

@aidenkeating @ciaranRoche do you think I should move that logic into the new snapshot providers?

Reply:
To save this already big PR from getting bigger I would vote to leave it as it is, as long as we can validate that snapshots created via the snapshot provider are tagged on reconcile of the postgres/redis provider. 

That work in the provider to tag resources, from my memory was added to make sure automatic snapshots taken by AWS had CRO tags on them. It was done before we has logic to take manual snapshots. So it might be worth a ticket (technical debt) to revisit this logic and flow if we are using it to tag both manual and automatic snapshots.  

What do you think?

Reply:
I'm +1 on that.

Reply:
A follow up ticket seems fine. If we are tagging the snapshots anyway I wouldn't see doing this as high priority.